### ISSUE TITLE
Bug: [no-unused-expressions] false negative on satisfies expression

### ISSUE THREAD

**Issue by Josh-Cena** [typescript-eslint/typescript-eslint#9275]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=5.4.5&fileType=.tsx&code=IwAgzghgLglmBmMCmYQDsCuBbARkgTgNwBQQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAm9AB7FoKZJQ5NUGSImjQO0SODABfEOqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
1 satisfies number;
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unused-expressions": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Error

### Actual Result

No error

### Additional Info

This was discovered while working on https://github.com/typescript-eslint/typescript-eslint/pull/9199. I will send a PR when that doc PR is merged.

**Comment by bradzacher** [comment]
My 2c - this shouldn't be matched.
A lone satisfies on top of a useless expression can be useful to ensure a type is what you think it is after refinements.

Eg `x satisfies string;` would enforce that your previous logic correctly set the variable such that no other types remained.
Or `x satisfies never;` would ensure that a branch is unreachable.

I do recognise that this is counter to the handling of `as` where we ignore the presence of the `as` and check its expression. But I personally believe it is intuitive.

Perhaps an option for this?

**Comment by Josh-Cena** [comment]
My personal intuition is that we should unconditionally report assertions and instantiation expressions and always allow `satisfies`, but with options to toggle each (`"ignore" | "report" | "transparent"`).

**Comment by guillaumebrunerie** [comment]
I don’t think this should be considered an error, either. I have used this pattern for unit testing generic types, for instance:
```ts
type MyGenericType<T> = T extends string ? true : false;

// Make sure it returns true for subtypes of 'string'
true satisfies MyGenericType<"abc">;
```

**Comment by Josh-Cena** [comment]
@guillaumebrunerie Why not use purely type-level operations? For example, this simple utility from [`@type-challenges/utils`](https://www.npmjs.com/package/@type-challenges/utils):

```ts
export type Expect<T extends true> = T
export type Equal<X, Y> =
  (<T>() => T extends X ? 1 : 2) extends
  (<T>() => T extends Y ? 1 : 2) ? true : false
```

You avoid confusing runtime expressions in your code.

**Comment by guillaumebrunerie** [comment]
@Josh-Cena I guess it might work as well, but sometimes it feels more natural to check that a specific runtime expression has the right type, rather than checking that the whole type is equal to another type. Here is another example:
```ts
type Types = {
    str: string,
    num: number,
};
type Kind = keyof Types;
export type Data<K extends Kind = Kind> = {
    kind: K;
    data: Types[K];
};

({ kind: "num", data: 3 }) satisfies Data<"num">;
({ kind: "str", data: "abc"}) satisfies Data<"str">;
// @ts-expect-error  Correctly reports an error.
({ kind: "num", data: "abc" }) satisfies Data<"num">;
// @ts-expect-error  Oops, wanted it to be an error but isn’t.
({ kind: "num", data: "abc" }) satisfies Data<"num" | "str">;
```

Also, it feels a lot more clunky to write
```ts
type _SomeUnusedNameThatIStillNeedToComeUpWith = Expect<Equal<MyGenericType<"abc">, true>>;
```
compared to 
```ts
true satisfies MyGenericType<"abc">
```

**Comment by Josh-Cena** [comment]
I can see this pattern being useful in a test file, but probably not sprinkled in the source file. Nevertheless I also agree that `satisfies` should not be reported by default, just that many times better alternatives exist. I think options to check/ignore them would be good though.

**Comment by dinofx** [comment]
```
1 satisfies number;
```

is transpiled to:

```
1;
```

This is clearly an unused expression (and odd use of `satisfies`).  This bug seems invalid.

**Comment by bradzacher** [comment]
> and odd use of satisfies

It's a pretty common one for type-system checking of things without runtime crashes

For example
```ts
switch (thing) {
  case 1: break;
  case 2: break;
  case 3: break;

  default:
    thing satisfies never;
    // enforces at the type-system level only that the switch is exhaustive.
} 
```

The alternative is writing a function that accepts an argument of type `never` - which is much heavier than just using the satisfies operator.

The expression is a no-op and will likely be just ignored by the engine during parsing. Any minifier run over your code would also likely cull the expression entirely.

**Comment by kirkwaiblinger** [comment]
> It's a pretty idiomatic one for type-system checking of things without runtime crashes

> The alternative is writing a function that accepts an argument of type `never` - which is much heavier than just using the satisfies operator.

This is a good tip! I've been doing `const __never_case: never = thing` but `satisfies` never occurred to me.

**Comment by dinofx** [comment]
Both the `default:` and the unused expression are unreachable/dead runtime code.  You can simply use https://typescript-eslint.io/rules/switch-exhaustiveness-check instead

**Comment by bradzacher** [comment]
> Both the default: and the unused expression are unreachable/dead runtime code

They may be. Until you add a new value to the union/enum and then they're not.
It's a convenient way to get that notification that you forgot to update a switch - especially if that switch is "non-terminal" (eg it doesn't have a `return` statement and isn't the last statement in a function body).

They also may not be truly unreachable eg if you're switching over server responses - the JS code may not be in sync with the server code. But if that's the usecase an explicit crash is usually preferred rather than an implicit no-op.

> You can simply use `switch-exhaustiveness-check` instead

If you use type-aware linting and want every switch to be exhaustive - then you're correct. But there are many codebases where one or both of these constraints does not hold true.

**Comment by kirkwaiblinger** [comment]
Just to give an explicit example of what the `never` trick does that may be preferable to relying on `switch-exhaustiveness-check`, you get to do this
```ts
type Thing = 'a' | 'b';

function useThing(thing: Thing) {
  switch (thing) {
    case 'a': 
       // do a thing
       break;
    case 'b':
       // do b thing
       break;
    default:
      // compile time error if 'c' is added to `Thing`; no runtime impact
      thing satisfies never;
      // alternately, what I've always done so far
      // const __never_thing: never = thing;

      // Handle this however you like if the types are incorrect at runtime
      console.error('types are incorrect at runtime!!');      
      break;

      // maybe you'd want to throw instead - it's up to you!
      // throw new Error(`unexpected thing: ${thing}`);
  }
}
```
[playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAKgFgSwHYHMoF4oHICGWoA+2ARlgNwBQFAZgK5IDGwCA9klLQM4TzIoAUwRKgBcsYSgCUUAN4UoUTgHcEwBnCiCJ0uQoUMc3bHjHy9egPQWoAExZQcUIXzPmoxAE4QcAa0rmDIyxSEVdzK1t7YicJML1Pbz8wmwhqHFoAG2BQtygIhhYAWzAEDOhmQugIDw8WDygEamwGfAROBxsUmyd7AANeVF6yKCR7D3oK6ARinCY4mL5FHGZOagQIdqQIADdq-zcInCzqpGWIDJAAGiglOGWoAEksXYcMpRwQdrstxXs0j3m+TYnGAUAA+mCtrsPGDnKIRjtqhgFqhKIDrAAJHBIGxlBbtOAsJSI+ogFi0KAZBA+KZNITlcAbBxeBqMOpeJgOUHjJCTeYFJCcFhlAB01VqHn4WFAkHaOBZyAKNQgnPuPMmAEINVhJMM3PMEr40bkIoUPsRoGTaFhuu9eT0YrUlKyQd5ugBaBrALDtWhgB1WjXox1EhHOgCiNTq-F69AgAA9IEwIN04SgxAASGRpgC+vV1ZhzFBzQA)

See also [this very pattern in the TypeScript docs](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#exhaustiveness-checking)

With `switch-exhaustiveness-check`, you need to leave the unexpected case _unhandled_ at runtime in order to get warned during linting. I personally prefer the flexibility to throw an error, or to log an error and move on with sensible default behavior, as the situation may call for, rather than being forced into... whatever may happen to come after the switch statement. 

Using `satisfies` for the `never` trick is new to me, but the `never` trick itself is well-established, and now that I'm aware of the use for `satisfies` there, I will probably switch to that. So this goes to say, I do see the value in at least having an option to exempt `satisfies` expressions from being considered unused.

**Comment by dinofx** [comment]
Regardless of preferred coding styles, this isn't really subjective.  `thing satisfies never;` is objectively an unused expression, no different than the [other 3](https://github.com/typescript-eslint/typescript-eslint/pull/8668).

It sounds like some folks would like to have some kind of syntax or utility type that does some kind of assertion during type checking, but otherwise has not runtime presence.  You might want to open/search for a TypeScript feature request in that case.

> I personally prefer the flexibility to throw an error

You could always do:

```ts
  default:
    throw new Error(`unexpected thing: ${thing satisfies never}`);
```

**Comment by Josh-Cena** [comment]
ESLint lints JavaScript. In JavaScript, an expression either has effects at runtime, or no effect at all. We lint TypeScript. In TypeScript, an expression can have effects at runtime, compile-time, or no effect at all.

Currently, we validate assertions and instantiation expressions by checking whether they have runtime effects, although arguably assertions have compile-time effects too. This is arguably _wrong_ because `a as number` is equally terrible as `a() as number`; you would only want to write `a()` in the latter case without `as number`.

It's not clear how many people rely on the compile-time effects of `satisfies`, but an option similar to ESLint's `allowTaggedTemplates` etc. would be useful, except we would be more nuanced and let users specify whether (a) it should always be considered unnecessary (b) it should always be considered necessary (c) it should be considered unnecessary if it has no runtime effects (current behavior).

**Comment by kirkwaiblinger** [comment]
@Josh-Cena 
I think we can proceed with the PR under this plan.

**Comment by dinofx** [comment]
> the other three that are unused both at runtime and at typechecking time

<img width="123" alt="image" src="https://github.com/typescript-eslint/typescript-eslint/assets/7581987/57b0b535-b419-4972-a8d7-b5df9ed9d0c4">

**Comment by bradzacher** [comment]
@dinofx yeah but try something like `3 as never` or `{} as {a:string}` and you'll get an unsafe result.

Which is why nobody does that ever and why instead people used to do `const _: never = 3` instead ([as shown in the TS docs](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#exhaustiveness-checking)).

**Comment by Josh-Cena** [comment]
Hello :) Now it's been 2 years, do we think if it should be a bug or a new option?

**Comment by bradzacher** [comment]
I think that an option is the correct behaviour.
There's many instances of people using `satisfies` as an easy-to-compile-away typecheck-only assertion.
The rule has behaved this way for a long time and [they implemented it the same way when they added support upstream](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC1leHByZXNzaW9uczogW1wiZXJyb3JcIl0gKi9cblxuMSBzYXRpc2ZpZXMgbnVtYmVyO1xuIiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7ImpzeCI6dHJ1ZX19LCJwYXJzZXIiOiJAdHlwZXNjcmlwdC1lc2xpbnQvcGFyc2VyIn19fQ==).

So this should be a new option to allow people to opt-in.

**Comment by Josh-Cena** [comment]
Oh right, in light of https://github.com/typescript-eslint/typescript-eslint/issues/11014, should we really work on these issues at all?

**Comment by bradzacher** [comment]
hmmm probably not yeah -- might be a good thing to file with core for their rule? Propose a new option to flag these?
